core/src/main/scala/cats/data/HashMap.scala

```diff
@@ -0,0 +1,1196 @@
+        Node.SizeMany
+      else
+        (valueElements: @annotation.switch) match {
+          case 0 => Node.SizeNone
+          case 1 => Node.SizeOne
+          case _ => Node.SizeMany
+        }
+    }
+  }
+
+  /**
+    * A CHAMP hash collision node. In the event that the hash codes of multiple elements collide,
+    * this node type is used to collect all of the colliding elements and implement the [[HashMap.Node]]
+    * interface at a performance cost compared with a [[HashMap.BitMapNode]].
+    *
+    * @tparam A the type of the elements contained in this node.
+    * @param collisionHash the hash value at which all of the contents of this node collide.
+    * @param contents the value elements whose hashes collide.
+    */
+  final private[HashMap] class CollisionNode[K, +V](
+    val collisionHash: Int,
+    val contents: Vector[(K, V)]
+  )(implicit hashKey: Hash[K])
+      extends Node[K, V] {
+
+    final def hasNodes: Boolean = false
+
+    final def hasValues: Boolean = true
+
+    final def allElements: Int = valueElements
+
+    final def valueElements: Int = contents.size
+
+    final def nodeElements: Int = 0
+
+    final def size: Int = contents.size
+
+    final def foreach[U](f: (K, V) => U): Unit =
+      contents.foreach(f.tupled)
+
+    final def contains(key: K, keyHash: Int, depth: Int): Boolean =
+      collisionHash == keyHash && contents.exists { case (k, _) => hashKey.eqv(key, k) }
+
+    final def get(key: K, keyHash: Int, depth: Int): Option[V] =
+      if (collisionHash != keyHash) None
+      else contents.collectFirst { case (k, v) if hashKey.eqv(key, k) => v }
+
+    final def getKey(index: Int): K =
+      contents(index)._1
+
+    final def getValue(index: Int): V =
+      contents(index)._2
+
+    final def getMapping(index: Int): (K, V) =
+      contents(index)
+
+    final def getNode(index: Int): Node[K, V] =
+      throw new IndexOutOfBoundsException("No sub-nodes present in hash-collision leaf node.")
+
+    final def add[VV >: V](newKey: K, newKeyHash: Int, newValue: VV, depth: Int): Node[K, VV] =
+      if (contains(newKey, newKeyHash, depth))
+        this
+      else
+        new CollisionNode(newKeyHash, contents :+ (newKey -> newValue))
+
+    final override def remove(key: K, keyHash: Int, depth: Int): Node[K, V] =
+      if (!contains(key, keyHash, depth))
+        this
+      else {
+        val newContents = contents.filterNot { case (k, _) => hashKey.eqv(key, k) }
+        if (newContents.size > 1)
+          new CollisionNode(collisionHash, newContents)
+        else {
+          // This is a singleton node so the depth doesn't matter;
+          // we only need to index into it to inline the value in our parent node
+          val mask = Node.maskFrom(collisionHash, 0)
+          val bitPos = Node.bitPosFrom(mask)
+          val newContentsArray = new Array[Any](Node.StrideLength * newContents.length)
+          var i = 0
+          while (i < newContents.length) {
```

Comment:
I think `foreach` may be better here since repeatedly accessing a vector is actually `N log N`, but foreach can be truly `O(N)`.